vertx-config-kubernetes-configmap/src/main/java/io/vertx/config/kubernetes/ConfigMapStore.java

```diff
@@ -0,0 +1,168 @@
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.json.JsonObject;
+import io.vertx.config.spi.ConfigStore;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * An implementation of configuration store reading config map from Kubernetes.
+ */
+public class ConfigMapStore implements ConfigStore {
+  private static final String KUBERNETES_NAMESPACE = System.getenv("KUBERNETES_NAMESPACE");
+  private final Vertx vertx;
+  private final JsonObject configuration;
+  private final String namespace;
+  private final String name;
+  private final String key;
+  private final boolean secret;
+  private KubernetesClient client;
+
+
+  public ConfigMapStore(Vertx vertx, JsonObject configuration) {
+    this.vertx = vertx;
+    this.configuration = configuration;
+    String ns = configuration.getString("namespace");
+    if (ns == null) {
+      if (KUBERNETES_NAMESPACE != null) {
+        ns = KUBERNETES_NAMESPACE;
+      } else {
+        ns = "default";
+      }
+    }
+    this.namespace = ns;
+    this.name = configuration.getString("name");
+    this.key = configuration.getString("key");
+    this.secret = configuration.getBoolean("secret", false);
+    Objects.requireNonNull(this.name);
+  }
+
+  /**
+   * For testing purpose only - inject the kubernetes client.
+   *
+   * @param client the client.
+   */
+  synchronized void setClient(KubernetesClient client) {
+    this.client = client;
+  }
+
+  @Override
+  public void close(Handler<Void> completionHandler) {
+    if (client != null) {
+      client.close();
+      client = null;
+    }
+  }
+
+  private Future<KubernetesClient> getClient() {
+    Future<KubernetesClient> result = Future.future();
+    String master = configuration.getString("master", KubernetesUtils.getDefaultKubernetesMasterUrl());
+    vertx.<KubernetesClient>executeBlocking(future -> {
+      String accountToken = configuration.getString("token");
+      if (accountToken == null) {
+        accountToken = KubernetesUtils.getTokenFromFile();
+      }
+
+      Config config = new ConfigBuilder().withOauthToken(accountToken).withMasterUrl(master).withTrustCerts(true)
+          .build();
+
+      DefaultKubernetesClient kubernetesClient;
+      try {
+        kubernetesClient = new DefaultKubernetesClient(config);
+        future.complete(kubernetesClient);
+      } catch (KubernetesClientException e) {
+        future.fail(e);
+      }
+    }, ar -> {
+      if (ar.failed()) {
+        result.fail(ar.cause());
+      } else {
+        this.client = ar.result();
+        result.complete(ar.result());
+      }
+    });
+    return result;
+  }
+
+  @Override
+  public void get(Handler<AsyncResult<Buffer>> completionHandler) {
+    Future<KubernetesClient> retrieveClient;
+    if (client == null) {
```

Comment:
should be synchronize to ensure client field visibility and that a single thread will set a Future on client